sql/expression/matchagainst.go

```diff
@@ -0,0 +1,343 @@
+	return &MatchAgainst{
+		Columns:          expr.Columns,
+		Expr:             expr.Expr,
+		SearchModifier:   expr.SearchModifier,
+		ftIndex:          expr.ftIndex,
+		KeyCols:          keyCols,
+		ParentTable:      parent,
+		ConfigTable:      config,
+		PositionTable:    position,
+		DocCountTable:    docCount,
+		GlobalCountTable: globalCount,
+		RowCountTable:    rowCount,
+	}
+}
+
+// GetIndex returns the relevant Full-Text index for this expression, or nil if it has not yet been set.
+func (expr *MatchAgainst) GetIndex() fulltext.Index {
+	return expr.ftIndex
+}
+
+// SetIndex sets the index for this expression. This does not create and return a new expression, which differs from the
+// "With" functions.
+func (expr *MatchAgainst) SetIndex(fulltextIndex fulltext.Index) {
+	if fulltextIndex == nil {
+		return
+	}
+	expr.ftIndex = fulltextIndex
+}
+
+// ColumnsAsGetFields returns the columns as *GetField expressions. If the columns have not yet been resolved, then this
+// returns a nil (empty) slice.
+func (expr *MatchAgainst) ColumnsAsGetFields() []*GetField {
+	var ok bool
+	fields := make([]*GetField, len(expr.Columns))
+	for i, col := range expr.Columns {
+		fields[i], ok = col.(*GetField)
+		if !ok {
+			return nil
+		}
+	}
+	return fields
+}
+
+// inNaturalLanguageMode calculates the result using "IN NATURAL LANGUAGE MODE" (default mode).
+func (expr *MatchAgainst) inNaturalLanguageMode(ctx *sql.Context, row sql.Row) (float32, error) {
+	var err error
+	expr.once.Do(func() {
+		// Evaluate the expression, which should always result in a string literal
+		words, nErr := expr.Expr.Eval(ctx, row)
+		if nErr != nil {
+			err = nErr
+			return
+		}
+		wordsStr, ok := words.(string)
+		if !ok {
+			err = fmt.Errorf("expected WORD to be a string, but had type `%T`", words)
+		}
+		expr.evaluatedString = wordsStr
+		// Grab the index for the doc count table
+		docCountIndexes, nErr := expr.DocCountTable.GetIndexes(ctx)
+		if nErr != nil {
+			err = nErr
+			return
+		}
+		if len(docCountIndexes) != 1 || docCountIndexes[0].ID() != "PRIMARY" {
+			err = fmt.Errorf("expected to find a primary key on the table `%s`", expr.DocCountTable.Name())
+		}
+		expr.docCountIndex = docCountIndexes[0]
+		// Grab the index for the global count table
+		globalCountIndexes, nErr := expr.GlobalCountTable.GetIndexes(ctx)
+		if nErr != nil {
+			err = nErr
+			return
+		}
+		if len(globalCountIndexes) != 1 || globalCountIndexes[0].ID() != "PRIMARY" {
+			err = fmt.Errorf("expected to find a primary key on the table `%s`", expr.GlobalCountTable.Name())
+		}
+		expr.globalCountIndex = globalCountIndexes[0]
+		// Create the parser now since it does a lot of preprocessing. We'll reset the iterators every call.
+		expr.parser, nErr = fulltext.NewDefaultParser(ctx, fulltext.GetCollationFromSchema(ctx, expr.DocCountTable.Schema()), wordsStr)
```

Comment:
Still a bit confused why the index and filter are both doing things with the parser/expr.Expr on the same object. We only need to parse the match expression once, right? I guess the reason is the second piece of functionality -- calculating relevancy counts? I would think the math expression mostly needs normalization, and the runtime lookups mostly need counting/frequency stats? Separating those sounds nice from a code clarity/organization/extensibility/testing standpoint. If I'm misreading/misunderstanding the intent also feel free to ignore my comments though.

Reply:
We do end up parsing the match string twice, but they're in different places. I mentioned it in another comment, but they're both interacting with the same tables to produce different results. The index wants to return the set of rows that contain the words present in the match string, while the expression calculates the relevancy of any given row. The index passes the rows to the expression, which then returns a relevancy which is interpreted as a bool (in the filter portion). The expression is not tied to the index and works just fine with full table scans, however the index is tied to the expression by relying on the fact that it has already been resolved by the point it is reached.

I _think_ that might answer your question?